### ISSUE TITLE
ASERT / NEFDA is the perfect DA?

### ISSUE THREAD

**Issue by zawy12** [zawy12/difficulty-algorithms#61]
This article is about the ASERT difficulty algorithm that was independently discovered by Jacob Eliosoff, [Mark Lundeberg](http://toom.im/files/da-asert.pdf), and a group at [Imperial College](https://arxiv.org/pdf/2006.03044.pdf). Jonathan Toomin did a lot of work ([here](https://read.cash/@jtoomim/bch-upgrade-proposal-use-asert-as-the-new-daa-1d875696) and [here](https://upgradespecs.bitcoincashnode.org/2020-11-15-asert/)) to show it's the best we have, to decide how long the crucial averaging window should be, and to implement it for BCH. 

There is a relative ASERT and an absolute ASERT. They are mathematically the same, but relative is expressed more like other difficulty algorithms and has a potentially large error if the "averaging window" is large due to nBit's having up to 2^(-15) imprecision and in approximating e^x in the necessary integer math. Absolute ASERT reduces those errors to effectively zero. EMA aka WTEMA is a very close approximation to relative ASERT.

**Relative ASERT:**
<img src="https://render.githubusercontent.com/render/math?math=\large target_{\text{i%2B1}} = target_i * e^{\frac{st}{TN} - \frac{1}{N}}">
```target[i+1] = target[i] * e^(st/T/N - 1/N)```
st = solvetime
target[i] = target at block height i, 2^256/difficulty
T = desired block time
N = mean_lifetime = half_life/ln(2)
The amount of error in difficulty and therefore solvetime is up to N * 2^(-15) as a fraction of the correct value.

**Absolute ASERT:**
<img src="https://render.githubusercontent.com/render/math?math=\large target_{\text{i%2B1}} = target_M * e^{\frac{t_i-t_M}{TN} - \frac{i-M}{N}}">
```target[i+1] = target[M] * e^((t[i] - t[M])/T/N - (i-M)/N) ```
t = timestamp
M = a fixed block height in the past.
The above is how Mark Lundeberg's [ASERT paper](http://toom.im/files/da-asert.pdf) expresses it, but to be perfectly correct, it needs to change target[M] to target[M+1].  The [BCH ASERT](https://upgradespecs.bitcoincashnode.org/2020-11-15-asert/) code expresses this correction as:
<img src="https://render.githubusercontent.com/render/math?math=\large target_{\text{i%2B1}} = target_M * e^{\frac{t_i-t_{M-1}}{TN} - \frac{i-M%2B1}{N}}">
```target[i+1] = target[M] * e^((t[i] - t[M-1])/T/N - (i-M+1)/N) ```

It's really interesting that the absolute form does not even need to depend on past targets or timestamps if you use the genesis time and just set difficulty to 1. This makes the absolute form:

target = e^((t/T-h)/N)
where 
h = current block height
t = current time since genesis (miners can just use their own local and with large N and reasonable tight validator limits, lying about current time will not lower difficulty unless miner wants other miners to orphan his block for cheating)

To make the weirdness more clear, if we use an "averaging window" of N=286 (BCH's is 288) and invert the absolute form to get it in terms of difficulty, ASERT is: 
**difficulty = 1.0035^E**
where
E = number of blocks the chain has emitted ahead of time.  So if genesis starts at D=1, it increases 0.35% for each block it gets behind.

**WTEMA algorithm:**
Relative ASERT is virtually identical to WTEMA by using the approximation that e^x = 1+x for small x. By expanding the parenthesis in the following integer math can be used, taking care to prevent target overflow or underflow (if negative solvetimes are allowed. ASERT can handle the negatives well). 
<img src="https://render.githubusercontent.com/render/math?math=\large target_{\text{i%2B1}} = target_i * \left( 1+ \frac{st}{TN} - \frac{1}{N} \right)">
```target[i+1] = target[i] * ( 1+ st/T/N - 1/N)```

## Problems

The first problem is that it gets behind or ahead in the number of blocks it emits over the long term as the hashrate increases or decreases by 
excess_blocks_emitted = mean_lifetime_in_blocks * ln(current_difficulty/start_difficulty)
All algos have this problem. The easiest and safest solution to reduce the error is to have a lower mean_lifetime less than say 144 blocks. This may come at a cost allowing switch mining to gain up up to 0.5% more blocks and makes it easier to conduct the "38% attack" described below.  If it's 400, there can be over 1% in excess block rate emission in normal conditions (and up to 38% in attack or attack-like conditions).  To make sure too many or two few blocks are emitted, the average solvetime can be checked over a longer averaging window and then adjust the target solve time from the official value by a percent that is in the opposite direction of the long-term error.  This turns out to be [a PI controller as described in a comment below.](https://github.com/zawy12/difficulty-algorithms/issues/61#issuecomment-664410510) The comments describes how to do it safely but the results will be worse if there is a long term oscillation that matches the secondary averaging window. If you choose a window that is too slow it has no effect.

Another problem is that if you use N=1, the median solvetime is ~ln(2)/2 instead of the expected ln(2) of the target time. This indicates it makes the difficulty too easy too often, despite being the only DA that has the perfectly correct average target solve time for N=1. If it's possible to have both, it might be evidence of having the perfect algorithm.

Another problem is that it can emit up to 38% more blocks in a long-term private mining attack than simply getting 100% of the blocks. This is done by setting the first timestamp in the private mine to 2x the mean lifetime window and then all the subsequent timestamps to 0 or 1 second. In SMA the attack gets 50% more blocks.  In LWMA I could not find any benefit to timestamp manipulation. If the average of say the past 6 solvetimes is used as the input for the "past solvetime" the attack can get 50% instead of 38%. If the max solvetime allowed is 6x block time, EMA can be reduced from 33% to 28%. It's not 38% because the exploit pushes the boundaries of its e^x approximation. 

Another problem is complexity of implementation. EMA is a lot simpler and almost the exact same so I prefer it it LWMA is not used.  N can't be small because large N allows the e^x=1+x approximation by making x small, but it can't be large because it has the average solvetime error like relative ASERT. I would use N=100 to 200. Also, it needs code I've mentioned elsewhere to handle out of sequence timestamps correctly.

Another issue (an easily solved "problem") is that the paper below recommends the absolute form of ASERT.  This exacerbates the 1st problem if a coin forks to ASERT and uses an initial target that is in the distant past: it immediately releases too many coins at the fork.  It will also be a problem if a coin changes the POW and needs to make a large sudden change to the difficulty. Both of these are easily solved by changing the baseline difficulty to the difficulty at the fork, but it's something to be aware of. 

Another issue is that **relative** ASERT and WTEMA accumulate error due to nBits which has 2^-23 to 2^-15 truncation error. From bitcoin wiki on difficulty: "0x008000 is the smallest legal value for the lower 24 bits since targets are always stored with the lowest possible exponent. "  This is only 2^(-15) = 0.003% error but it decreases average solvetime in proportion to N. The max decrease (error) in solvetime is approximately 2^(-15) * N = 1.8% for N=600. 

The error in the e^x approximation in the code accumulates in the same way in **relative** ASERT (and WTEMA), requiring higher precision in the integer-based e^x estimator if N is large. The error might be made to offset the average nBits error.

A trivial fix for the last two problems is to change the target solvetime by the same amount to offset them, but an attacker doing a private mine (or miners as a group) could select timestamps that minimize difficulty by about 1/2 the error for which you correct. These last two problems are why absolute ASERT is usually recommended. But N=100 is a good setting for relative ASERT and has only 0.3% error in the nBits portion.

### My "Derivations" of ASERT

For **Relative ASERT**, there seems to be direct & solid argument in claiming it's simply using the exponential distribution:  (I've inverted the equation to use difficulty instead of target).
D = prev_D * exp_dist_observed / exp_dist_expected
D = prev_D * [ 1/T * e^(-st/T) ] / [ 1/T * e^(-T/T) ]
D = prev_D * e^(-st/T) / e^(-1)

This can vary a lot with st values jumping around, so a smoothing factor is needed. Since a sequence of adjustments to D are multiplying the ratio over and over, an Nth root of this equation can be used. In physics equations the N in e^(-x/N) is called an attenuation depth or mean free path.  The best name in this context is "[mean lifetime](https://en.wikipedia.org/wiki/Exponential_decay)" where time is measured in blocks. An SMA has the same Std Dev of difficulties if its averaging window is 2N. 

Including a mean lifetime low-pass filter to what I have above gives relative ASERT:

D = prev_D * e^( (1-st/T) / N).  

**Absolute ASERT** is mathematically the same as relative, but its form allows a different explanation. Apparently, we can claim it's a solution to the differential equation that says we should increase the difficulty per error E (the number of blocks behind schedule) in proportion to the current difficulty. 
dD(E)/dE = k\*D(E)
The solution to this is
D(E) = D0 * e^(k\*E)
which is absolute ASERT with k=1/N.

## More complicated derivation

**Only previous block data is needed**
To estimate current hashrate H we use past difficulties (D's) and solvetimes (st's). If our previous estimate was the best possible given all past D's and st's, then just the previous D and st should be sufficient to modify our previous H estimate. Even if we did not use all past D's and st's and our estimate is far from perfect, a perfect correction should be in the right direction and trend towards being perfect, so over the long term it will actually be perfect, so we only need the "single-block" adjustment to be perfect. Given an H we expected in the prior block, we want to adjust D like this:

D = prev_D * H_actual / H_expected

We can't use solvetime to directly estimate H_actual because a solvetime that is 10% high or low does not indicate H is 10% low or high. This is because solvetimes have a probability distribution around T that will persistently overestimate H because solvetimes are much more frequently < T than > T.  We need to map the relationship between st and H to some linear function that properly estimates H with only a single st. The relationship between H and st is given by how we model solvetimes:

CDF = 1 - e^(-st * H / D)
1 - CDF = e^(-st * H / D)
CDF & 1-CDF are random variables from 0 to 1. All equal-sized ranges within 0 to 1 have equal probability. I use the 2nd equation to model solvetimes with 1-CDF = x = rand(0,1).
x = e^(-st * H / D)

If avg(st) over many blocks = T, the D was selected correctly for the given H, i.e. H / D = 1/T so that on average we will see:
x = e^(- avg(st) /T) = e^(-1).
If H is 10% higher than expected, i.e. H_actual = 1.1 * H_expected, then avg(st) will be T/1.1 too low for the D. But we have to use the x equation to estimate avg(st) from a single st. 

I'll claim the random variable "x" is the desired linear mapping of the solvetimes that takes the probabilities into account.  I can't explain/prove it, but it seemed apparent when I first saw ASERT / NEFDA that this is what it's doing:

D = prev_D * x_actual / x_expected
D = prev_D * e^(-st/T) / e^(-1)

**Comments on the Imperial College Paper**

Below is a review of a new difficulty algorithm paper, "[Unstable Throughput: When the Difficulty Algorithm Breaks](https://arxiv.org/pdf/2006.03044.pdf)" .  ( Their tn in this will be the timestamp of prev_D, not current time. )

IMO the paper is good, except for these notes / complaints. I appreciate the author's citing my DA articles and speaking well of this repository of work I've put into it. As with Tom Harding's and Mark Lundeburg's papers, it's good to see better mathematicians carry things through.

For the sake of history in trying to understand why it's taking so long to find the best DA, Jacob Eliosoff is the first one I know of who considered this algorithm, knowing it may have been the best, but opted for the approximation (EMA) due to wanting to avoid the exponential. 

**Section 3.2**
I believe BCH very closely followed the Poisson distribution for a year post-EDA when on-off mining was not a problem due to a higher price & subsidized mining.  The depictions of how far away from the ideal Poisson seem a little unfair in the sense that problem happens only when on-off mining is a big problem. It's infinitely better than BTC's. Further, a coin has a much bigger security issue than the DA if on-off mining is a problem because it shows > 50% of miners do not depend exclusively on the coin for income which is the sole basis of BTC's security.  

SMA's do not directly cause oscillations as a result of high or low solvetimes falling out the back of the window because new solvetimes coming in the front and the entire averaging should balance out.  The SMA is a fair estimate of the current hashrate as it was in the past at 1/2 the size of the averaging window. Accidentally going low with accidental slow solvetimes out the back or long solvetimes in the front can motivate extra miners to come online. So in that sense it can be a "cause", but it doesn't oscillate if H is constant.  Once miner motivation causes a significant change for several blocks, the SMA does not dampen the oscillation. It invites a continuance.  A close analogy: a steel bar does not ring with oscillations unless you strike it while it's in the air (no dampening from holding it). The strike causes the ringing, not the bar. 

**Oscillations made worse by delays**
The median of 3 blocks BCH uses may directly cause a very slight oscillation due to the delay. Delays like using the MTP as the most recent st as the authors suggest (and as is done in Digishield) may directly cause some oscillation. Zcash has always had a mild oscillation from using MTP. Monero's algorithm caused huge problems in alt coins that copied it because of the various ways it causes a delay (does not use the most recent solvetimes).  All alts I'm aware of that used it "blew up" and got stuck until they switched algorithms.

**Section 4.2.1**
D/st is not an estimate of H for small samples due to increased frequency of small st values making avg(1/st) > 1/avg(st) = 1/T.  The best H estimator I have is sum(D)/sum(st) * (N-1)/N.  I do not know how to get hashrate or the expected number of hashes from 1 solvetime unless we also have an a priori expected value of the solvetime (i.e. H), in which case it appears the "observed" H for 1 block is D/T * e^(-st/T) where D/T is the expected H. 
I wonder if there is an e^x function that can do away with the (N-1)/N and always correctly estimate HR. BTW If solvetimes are prevented from being < 1, the estimate of H for 1 block that gives the right average is: avg(1/st)  = D/st/7. The avg H estimate for 1 block is not an integral that converges if st can be 0.

**Section 4.6**
I've never seen any argument against having a future time limit of say 10 seconds. Based on distributive consensus theory, it only needs to be > 2x propagation delays and should be << block solve time.  Block solve time is purposefully chosen to be >> propagation delay. There is no BFT without a pre-consensus on time and its limits.  A 2-hour or 10 second rule are not honored because >50% of miners decide to. It's honored because each honest node will immediately reject a chain that is ahead of time due to world time being an objective and obvious fact that the protocol and network do not, should not, and can't provide.  Peer time is not the basis because it's rejected if it is 70 minutes off from local time. It should not have been used.

**Section 5.5**
NEFDA-MTP-72 (which is has the same smoothness as BCH under constant H) is as bad as BCH's DA in terms of delays & "stolen" blocks when modeling on-off mining problems. It invites bad oscillations. The oscillations are faster but more frequent.  My method of testing it is a step function that goes on or off if D goes too low or too high, respectively. Since a 10% change in D can cause a 10,000% change in HR, it's pretty accurate.  I've used this a lot to simulate live attacks very accurately. 

![image](https://user-images.githubusercontent.com/18004719/84327017-4b745200-ab4c-11ea-8df6-e9783028885e.png)

**Comment by dii14** [comment]
Hello, 

Happy to see you reviewed the paper so promptly.

Your comments / complaints all seem valid points. Let me respond to them in order:

**Section 3.2**
We wew not trying to say that BCH is really bad just that the situation can get much worse, maybe we were not very explicit. You are absolutely right saying that the situation used to be better and this can be seen from the plot measuring number of blocks per hour, but as you rightly point out this only used to happen when coin-hopping was not happening and mining was subsidised. I think in the future coin-hopping will happen more and more as mining pools introduce auto-switching functionalities, so deserts and spikes will indeed get out of hand.

I think the problem with a sliding window is actually an important factor for sustained cyclicality. It's just not right that a solve time influences the hash rate estimation twice: when it enters the window with negative feedback (i.e. if the solve time is long, it increases the window length and causes the difficulty to go down) and also when it falls off the window with positive feedback (i.e. when a long solve time falls off the window, its length decreases and the difficulty goes goes up). The reason I don't believe the entire average is balancing out is that this is not just a random process where long block times occur randomly because miners will respond to these oscillations by mining according to profitability. Positive feedback should not exist in any DA because it creates incentives that should not be there. This positive feedback is essentially sustained by coin-hoppers and that's how the cycle of long block times followed by short solve times forms. The fact that the SMA also has a delay, is a major factor as well as you describe very well in some of your posts, because it allows coin-hoppers to operate without the DA noticing their activity immediately. I think we basically have the same views, just that we believe that "miners causing the oscillation" is actually the DA's fault for creating the incentive in the first place. 

4.2.1
"D/st is not an estimate of H for small samples due to increased frequency of small st values making avg(1/st) > 1/avg(st) = 1/T." This is indeed absolutely true. We explain this issue when we argue for the introduction of the correction from 1 + T/S to e^(T/S). The reason this is not just an approximation (like it usually is for small values of T/S) but actually a correction is exactly what you describe. We are not quite sure how to include the effects of more probable shorter st without getting some very convoluted probabilistic formulas so instead we gave that indirect argument relying on a scenario under which the algorithm must work correctly.

4.6
"A 2-hour or 10 second rule are not honored because >50% of miners decide to. It's honored because each honest node will immediately reject a chain that is ahead of time due to world time being an objective and obvious fact that the protocol and network do not, should not, and can't provide." We were trying to make 2 points here and I think we are mostly in agreement over them. First of all this rule is not a consensus rule as miners catching up with the network cannot impose it at all. Secondly, it is not at all obvious what constitutes a block that lies about its timestamp by placing it in the future. If miners would not globally agree on a certain cutoff time that is approximately the same for all of them (be it 2hrs or 10 mins in the future) then they would not know which blocks to consider as adopted by others as well and start mining on them or ignore them as the block is too much in the future and keep mining on their own block. If there would not be majority consensus over the exact time a block is allowed to be in the future (I don't understand why 2 hours it seems way too much to me) then miners would essentially not know what blocks their peers consider as valid at this time and might keep mining on the previous chain hoping others do so similarly. Essentially this would lead to many different competing forks that miners switch between as they grow and as blocks with various future timestamps are appended.
The reason we make the devil's advocate regarding breaking this 2-hour rule is to argue that using RTT is actually a viable alternative because the rules is not just some rule that miners *happen* to respect, but in fact everything would descend into chaos without it as miners would not be able to agree what amounts to lying and what not. So a majority of nodes do indeed need to agree on how much can a new block's timestamp differ from the current time.

5.5
Indeed, NEFDA-MTP-72 is really quite bad. We did not want to create the impression it's good at all. We were just trying to make the point that if the smoothing factor is not large in comparison to the MTP window size then this leads to similar oscillations problems. We were giving this configuration as a bad example, and for showing how to mitigate its problems by increasing the smoothing factor. We really like your simulations with stolen blocks but the space limitations of an academic paper kind of limited the depth of analysis we could go into. In my opinion simulating miner's behaviour is a whole paper on its own. In fact the model section of our paper could probably be more rigorously extended by proving the properties it has in a short paper as well.

Overall it seems like we are on the same page, which is very much expected as we studied your work a lot and found it really detailed, rigorous and insightful.

**Comment by zawy12** [comment]
We agree SMA's plus miner motivation is a positive feedback situation. But then every DA that is not an RTT causes positive feedback. A major reason for having a long-window SMA is to minimize the variation that would affect miner motivation, so short solvetimes rolling out the back have little effect and are (usually) soon counteracted by long solvetimes. In short, a delay in estimating current hashrate plus miner motivation causes oscillations.

No one has ever stated a reason for 2 hours except for irrational fear over daylight savings time. I've discussed it with a lot with small coin devs and no one could think of a reason that 10 seconds would not work. Lamport's 1978 paper on clocks and his byzantine paper give the requirements for clocks in order to reach distributed consensus. If these had been followed, there would have never been a timestamp attack of any type.  One rule is that future time limit needs to be significantly longer than propagation delays and a lot less than block time because block time is what's measuring the number of voters who have purchased their lottery ticket (each hash).  Another rule is that timestamps must be sequential (no MTP needed). 

My TSA article is all about RTTs and discusses effect of timestamps in detail. There are a lot of complex things. My conclusion is that the future time limit (FTL) should be very tight and that difficulty should not be affected hardly any by timestamps set at that limit.  So my simple RTT at the top of the TSA article is what I believe is the best DA: a fast mild RTT riding on top of a slower DA.  I decided chain work should be based on the slow DA w/o the RTT adjustment.  If not, then lying to get an earlier timestamp to win chain work might occur, complicating things, possibly allowing an imaginative attack.  If an RTT add-on does not affect chain work, then everyone may use a forwarded time if you don't have a tight timestamp. They would all agree to "push time to the FTL" constantly having timestamps that fall on top of the FTL, to the extent they don't think their blocks would be rejected for being too far forward. This could cause forks, which they don't want, so they would find a happy medium but always still pushing to closer to the FTL, so you might has well have nodes enforce it closer just so that time is more accurate.

Something to keep in mind for competing tips is that chain work is not sum(D), but  
sum(D) * (N-1)/N
Emin et al have 2 articles that discuss successful attacks with around 35% HR and I wonder if it's because chain work is not being correctly measured. I have not read their [newer paper](https://fc20.ifca.ai/preproceedings/4.pdf) yet.  I keep looking for an e^x calculation that can be more accurate for estimating HR and chain work for a small number of blocks.

**Comment by dii14** [comment]
Well the point is to have less positive feedback than negative feedback. That's why lwma and ewma/asert/nefda work well. Because the blocks that are in the past are weighted less so the positive feedback they amount to is diminished in time. The problem with SMA is both the fixed window and the equal weights.

I completely agree on the 2-hour rule. No idea why it is so long and I agree it could be diminished. I think the daylight saving time is a really weak argument nowadays. I think it's a good proposal to keep it shorter than ideal inter block time. We just have to make sure miners use good clocks but I think this is expected with current computers.

I'm not sure I understand why you would lie with an earlier timestamp if the algorithm is RTT. First of all you have to set the timestamp before starting to hash. And if you succeed to find a solution before that lower timestamp then that just means you really did do much work. Am I misunderstanding you? About tight FTL, absolutely, that's a very good reason to have short FTL. Even with short FTL I don't think we can ever get any of these blockchains to be a correct time-stamping server like the BTC whitepaper suggested so even if the blockchain time is not correct at all that's not such a serious problem in my opinion.

Good point about chain work! and thanks for the link, I'll look into it.

**Comment by zawy12** [comment]
My conjecture that this is the perfect DA seems wrong (although it could still be in some sense the best possible). To fill in the gap where I hand-waved above, recall H and st are related by 
x = rand(0,1) = e^(-st * H/D)
which means
st = -ln(x) * D / H 

Let k =  fraction of err in H_actual/H_expected
k = H_actual/H_expected - 1
i.e. k=0.1 means H is 10% higher than expected and thereby needs a 10% increase in D. We want a DA that does this:
D = prev_D * (1+k)

Using the eq above for st, our DA (without the smoothing factor) in terms of x is
D = prev_D * e^(ln(x)/(1+k) +1)
x = rand(0,1) is is equally probable in any range, so the expected (average) adjustment from our DA is
![image](https://user-images.githubusercontent.com/18004719/84370843-d2561880-aba6-11ea-8d4b-1f6276ab48d6.png)
where Wolfram Alpha is a math chauvinist for saying log() = ln().

In other words, our average correction is (after rearranging the integral result):

D = prev_D * (1+k) / (1+k/2) * e/2

So it's not the (1+k) I hoped for. 
I separated the e/2 because that's what it is when k=0. So if no adjustment on average is needed, it's going to adjust by 
D = prev_D * e/2

With the smoothing factor of M blocks the adjustment to difficulty is 
![image](https://user-images.githubusercontent.com/18004719/84380872-1b619900-abb6-11ea-81dd-304cb8c5e34c.png)

**Comment by zawy12** [comment]
On average, ASERT / NEFDA makes no change to D if D = H * T * (e-1). Every adjustment is pushing D towards this. But due to the PDF of st, it is constantly over- and under-shooting this target such that the average D = H * T.

Proof, using same method as above:
![image](https://user-images.githubusercontent.com/18004719/84506215-04da4100-ac8d-11ea-8a9b-9022bf23142e.png)

It might be viewable like this:

D = prev_D * exp_distr_observed / exp_distr_expected

D = prev_D * (1/T * e^(-st/T) ) / (1/T * e^(-T/T) )

**Comment by zawy12** [comment]
I changed my chain work article to include ASERT / NEFDA as the estimator.  It gives much more reasonable results that using sum(D)/sum(st) * (N-1)/N.  In short, I use this DA to estimate the number of hashes by assuming a 2-tip chain split has 50% of the hashrate for each of the tips. This gives the tips an equal opportunity to win.

**Comment by jacob-eliosoff** [comment]
@dii14, and any others reading along - just FYI here are a couple of old threads on our old EMA DAAs that I reread when I'm trying to remind myself how they work.  The specific algos (@dgenr8's `wtema`, my `ema`/`ema2`) aren't quite identical to ASERT/NEFDA (the one that is is my `simpexp`), but they have the same exponential-filter spirit and `wtema` is very elegantly implemented.

- Introduction to `ema`: https://github.com/kyuupichan/difficulty/pull/26#issuecomment-342290398
- Discussion of `wtema`: https://github.com/kyuupichan/difficulty/pull/30#issuecomment-355854001

**Comment by zawy12** [comment]
The best way to view wtema is to simplify the math:
next_target = prev_target*(1 + st/T/N -1/N)
next_diff = prev_diff / (1+st/T/N - 1/N)

For largish N, it's same as ASERT/NEFDA
next_diff = prev_diff * e^(-st/T/N + 1/N)
N = mean lifetime in blocks.
Std Dev in diff =~ 1/SQRT(2 * N) 

ETH intuited an approximate inverse wtema via 1/(1-x) =~ 1+x.
next_diff = prev_diff * (1 - st/T/N + 1/N)

They saw the potential problem of a negative difficulty for large st, so they used a max() function to stop it.  We rejected this version because of that potential.  Their max() function has a drawback: it enables a 50% selfish miner to get I believe 174% of blocks in the time they would normally get 100%.

Digishield and Grin are an SMA inside a wtema: 
next_target = avg_17_targets* (1 + avg_17_st/T/N - 1/N)
N = 4 Digi, N=2 Grin. 

There is still the possibility of negative difficulty with wtema in BCH if using wtema is the only change made. It can happen if an attacker gets 2 blocks in a row to get around Amaury's median of 3 and sets his timestamps to the MTP+1 and if the previous 6 blocks were so slow such that N is not sufficiently large.  For example N=72 and last 6 blocks took an average >12xT it could happen.

**Comment by tromp** [comment]
> Digishield and Grin are an SMA inside a wtema:
> next_target = avg_17_targets* (1 + avg_17_st/T/N - 1/N)
> N = 4 Digi, N=2 Grin.

Grin uses a DIFFICULTY_ADJUST_WINDOW of  HOUR_HEIGHT  = 60 
and DIFFICULTY_DAMP_FACTOR = 3
The DAA function consensus:: next_difficulty is at
https://github.com/mimblewimble/grin/blob/master/core/src/consensus.rs#L309-L356

Note that it adjusts average difficulty, not average target.
Grin enforces strictly increasing timestamps.
The difficulty chart at https://grinscan.net/charts shows there's still a fair bit of oscillation.

The main reason for preferring damped SMA over something like wtema is worry about timestamp manipulation; a sum of the past 60 solvetimes seems much harder to manipulate than just the last one.

**Comment by zawy12** [comment]
Since it uses difficulty, replace my avg target should be the harmonic mean.  Plug in N=2 and you should then see the equation I have is the same as yours.

**Comment by tromp** [comment]
Are you saying N corresponds more to a dampening factor than to a block window?

**Comment by zawy12** [comment]
N is the mean lifetime = ML in e^(-t/ML).  The best alternate description might be calling it a filtering constant as jonathan toomim found [this paper](http://www.tsdconseil.fr/tutos/tuto-iir1-en.pdf) that shows EMA is a first order IIR filter.   If ASERT is perfect and Grin is an approximation, then perfect Grin is

difficulty = avg_difficulty * e^(-avg_st/T/2 + 1/2)

**Comment by zawy12** [comment]
Using e^x = 1/e^-x = 1/(1-x) for small x gives one of the alternate EMA formulas. Then multiplying by N/N and T/T gives me:
difficulty = avg_difficulty * N * T / ((N-1) * T + avg_st/T)
OK, so it seems my N is your damp factor.  N=3.  I was just remembering wrong when I wrote N=2.

**Comment by tromp** [comment]
In wtema, you don't compute any average target or difficulty; you just multiply the previous one by some factor depending on last solvetime to get the next.
Multiplying the target by X is exactly equivalent to dividing the difficulty by X.

I prefer to use N to denote the half-live in blocks,
and then set alpha  = 1 - 0.5^{1/N}, leading to a wtema update of

  next_diff = prev_diff / (1 + alpha * (st/T - 1))

**Comment by zawy12** [comment]
In my "perfect"  grin above, the x in the above is often not small, so I'm not at all sure the "perfected Grin" is good. 
 
Half-life is easier for people to understand, but I want to keep things in mean_lifetime to use the e^x approximation. 
 half_life = ln(2) * mean_lifetime.  Alpha keeps it connected to EMA notation.  Another benefit is that SMA and LWMA's mean lifetimes in terms of equivalent std dev of difficulties with ASERT/EMA is 1/2 their N.  LWMA N=200 has its closest connection to ASERT/EMA mean_lifetime 100.  Grin's mean lifetime should be 3x the averaging window, also divided by two.  I know this from experiment with Digishield.

**Comment by tromp** [comment]
> Grin's mean lifetime should be 3x the averaging window, also divided by two.

Aha, so when I said

> What choice of N would give an overall behaviour closet to Grin's DAA? Would it be 60 * 3 ?

I was only off by the division by 2, and N should be 60  * 3 / 2 = 90 ?

**Comment by zawy12** [comment]
Yes, if you're averaging 60 blocks and "lifetime" units are in blocks as opposed to seconds.  The equivalent std devs are with constant hashrate. I have a theory that the expected std dev in exchange price should be about 2x the std dev of difficulty under constant hashrate, as determined by mean lifetime.

**Comment by tromp** [comment]
Thanks for elaborating. I should do some simulations for Grin's current DAA vs this simpler twema alternative to evaluate the pros and cons.

**Comment by zawy12** [comment]
WTEMA is going to be better in all respects because grin is partially SMA.  To make the DA's work better and prevent exploits, make FTL < block_time (at least no more than 3x block time) and especially important to prevent all kinds of hassles is to make the consensus rule on timestamps > 0 from previous timestamp (override local time if necessary to make it +1 if previous block as long as it is < FTL).  If you use an MTP(11) =~ 6 blocks in past for the solvetime, make sure to use prior difficulty, not the one in the past (you should be doing that now with your difficulty average) and also make the mean lifetime at least 20x the 6 blocks or it is asking for oscillations. Another option to deal with negatives and allow WTEMA to use most recent solvetime is to do the "solvetime conditioning" loop I have in LWMA for the MTP=11 blocks (going back 12 blocks to get initial timestamp). Allowing negative solvetimes or dealing with them in a different way than these 3 options is fraught with problems in all DAs.

**Comment by zawy12** [comment]
@dii14 ASERT/NEFDA is going to be used for BCH.  It emits too many or too few blocks if hashrate consistently increases or decreases.  The excess blocks it emits is M * ln(current_difficulty/start_difficulty).  To correct for this, we make the block time that is input to ASERT a function of the long term solvetime error. It can actually be an absolute ASERT to adjust block time inside an ASERT that adjusts target, but we use the e^x = 1+x approximation.   BCH will use a relative ASERT for the target.  

Relative ASERT:
h = height
T=desired block time
t = timestamps
M = mean lifetime in blocks if each block is T seconds. M*T = mean lifetime in seconds.
k = additional dampening factor
target[h+1] = target[h] * e^( (t[h]-t[h-1] - T[h])/M/T)
T[h] = T * max(1/1.13, min(1.13, 1 - ((t[h]-t[0]) + T*h)/M/T/k ) )
The 1.13 clamp prevents a private mine attack from getting a much higher number of blocks than the usual 100% per time period.  I am recommending M=144 and k=5. BCH will probably use M=416  (a halflife of 416 * ln(2) = 288) and maybe a different k. 
Using e^x = 1+x approximation give the DA algorithm I'll recommend to people to avoid the e^x complexity.  The truncation error is very small as I described in an update above.
target[h+1] = target[h] * (1 + (A - B) + B*(C-D)/5)
A = (t[h] - t[h-1] ) / 144 / T  = prior observed block time in units of blocks, filtered
B = 1/144 = previous block expected time, filtered
C = (t[h] - t[0]) / 144 / T = past blocks observed, filtered
D = h/144 = past blocks expected, filtered

This is a typical PI controller of the form
next = prior * ( J + K * x + L * sum(x<sub>i</sub>))
with constants 
J = offset
K = slope
L = undershoot
and x = error signal = observed solvetime minus desired solvetime.

ASERT just gave us a good handle on choosing the constants.

**Comment by zawy12** [comment]
This is the derivation of how nBits trunction error E as a fraction of nBits accumulates in relative ASERT to cause avg solvetime error to be E * H where H = mean_lifetime.  For example, the max nBits error in a certain range of nBits is 2^(-15) for half life 288 this is 2^(-15) * 288/ln(2) = 1.3%

This will make extensive use of small x approximations:
e^x = 1+x
x+x^2 = x 
/(1-x) =~ 1+x 

Relative ASERT with small nBits truncation error E:
target[h+1] = target[h] * (1 - E) * e(^(t[h] - t[h-1] - T)/T/H)
T = target block time
h = height
t[i] = timestamp
H = mean lifetime

If target[h] was correct for the current hashrate then target[h] = ideal_target and the solvetime on average was T. But because of the E, the target[h+1] will be too small: 
target[h+1] = ideal_target * (1-E) * e^( (T - T)/T/H) 
target[h+1] = ideal_target * (1-E)

The next block will not have an avg solvetime of T, but of T/(1-E) due to the harder target.
The next block will be:
target[h+2] = target[h+1] * (1-E) * e^( (T/(1-E) - T)/T/H) 
target[h+2] =~ target[h+1] * (1-E) * e^( E/H)   # using 1/(1-E) = 1+E
target[h+2] =~ target[h+1] * (1 - E + E/H )   # using 2 approximations

The accumulated error so far using target[h+1] = ideal_target * (1-E) is:
target[h+2] = ideal_target  * (1-E) * (1- E + E/H) 
target[h+2] =~ ideal_target * (1- 2E + E/H) 

The next block will have:
target[h+3] = ideal_target * (1- 2E + E/H)  * (1-E) * e^( (T/((1- 2E +  E/H) - T)/T/H)
target[h+3] = ideal_target * (1- 2E +  E/H) * (1-E) * e^( (+2E - E/H)/H)
target[h+3] =~ ideal_target * (1- 3E +  E/H) * (1+2E/H - E/H^2) 
target[h+3] =~ ideal_target * (1- 3E + 3E/H)  

Next block:
 target[h+4] = ideal_target * (1- 3E + 3E/H) * (1-E) * e^( (T/(1- 3E + 3E/H) - T)/T/H)
 target[h+4] =~ ideal_target * (1- 4E + 3E/H) * (1+3E/H)
 target[h+4] =~ ideal_target * (1- 4E + 6E/H)

 The ultimate pattern is 
 target[h+N] =~ ideal_target * (1-NE+(N-1)N/2*E/H) 

At a certain point, the current target correction downward at N will equal the inverse of the correction at N+1 as a result of the solvetime equaling what's needed.  Theoretically, target will oscillate between those two corrections. 
(correction at N) = 1/(correction at N+1)
(1-N*E+(N-1)*N/2*E/H) = 1 / (1-(N+1)*E+N*(N+1)/2*E/H)
Ignoring all the square terms:
1 = 1-NE+(N-1)N/2*E/H - (N+1)E+N(N+1)/2*E/H
H = N - 1 - 1/N =~ N

Sub H for N in the correction at N to get the error in solvetime
1/(1-NE+(N-1)N/2*E/H)  = 1/1-E(N+(N-1)/2) =~ 1/(1-E(H-H/2)) = 1+EH/2

But EH/2 is the error in solvetime to get a stable target that is not the ideal_target for the hashrate. The target is lower than desired by this same amount which causes solvetime to be that much longer.  Therefore delays = EH.

**Comment by zawy12** [comment]
People forget that ASERT gets behind in the number of blocks if hashrate continually increases because it has perfect solvetime if hashrate is oscillating around some average. Getting behind is how it increases difficulty.

This is the derivation of how many blocks ASERT will get ahead or behind as difficulty increases or decreases.
T = desired block time
t[h] = timestamp at height h
H = half life in time units of T (i.e. the number of blocks *if* they were all emitted in T seconds)
T*H = half life in seconds
Absolute ASERT:
target[N] = target[M] * 2^( ( t[N] - t[M-1] - T*(N-M+1) )/T/H)

Average solvetime is T*(1+F) where F is the fraction of delays.
target[N] = target[M] * 2^( ( T*(1+F)*(N-M) - T*(N-M+1) )/T/H)
Solve for F*(N-M) = number of blocks behind
 F*(N-M) =  288 * log( target[M] / target[N] )/log(2) - 1 
 F*(N-M) = 288 * log(difficulty[N]/difficulty[M])/log(2) -1

**Example:** 
BTC has had about 6% faster solvetime than desired due to hashrate increasing faster than the DA.  What would BCH's avg solvetime have been if it had used ASERT with half life 288 from the beginning?  Difficulty is 350E9 and it started at D=1. The emission schedule wanted about 605,000 blocks emitted by now instead of 640,000 (5.7% too fast).  

288 * log(350E9)/ln(2) / 605,000 = 1.8% too fast (too many blocks)

**Comment by tromp** [comment]
I forked jtoomim's comparison tool in order to compare Grin's current DAA against some other damping factors and against wtema with various halving times.
Since Grin does't use nBits, I stripped it out, and just store precise targets in the states instead. See

https://github.com/tromp/difficulty

**Comment by tromp** [comment]
dear Scott,

would you care to comment on my proposal for upgrading Grin's DAA to wtema,
at
 https://github.com/mimblewimble/grin-rfcs/pull/61

?

Thanks.

regards,
-John

>

**Comment by A60AB5450353F40E** [comment]
Here's something interesting.

Recall the "absolute" version:

`target_{n+1} = targetAtActivation * 2^((timestamp_n - targetTimestamp) / tau)` ,
where `targetTimestamp_{n+1} = (n + 1 - n_activation) * targetBlockInterval` .

From there, we can simply calculate relative (block-to-block) adjustment by dividing some block's difficulty with the difficulty of block before:

`target_{n+2} / target_{n+1} = (targetAtActivation * 2^((timestamp_{n+1} - targetTimestamp_{n+2}) / tau)) / (targetAtActivation * 2^((timestamp_n - targetTimestamp_{n+1}) / tau))` .

The constants cancel each other out, the 2 `timestamp`s reduce to `solvetime`, and 2 `targetTimestamp`s reduce to `targetBlockInterval`, and we get:

`target_{n+2} / target_{n+1} = 2^((solvetime - targetBlockInterval) / tau)` ,

or what you call the "relative" version.

Mathematically speaking, the "relative" is the [finite difference derivative](https://en.wikipedia.org/wiki/Finite_difference_method) of the "absolute".
Or, looking at it from the other direction, the "absolute" is the [closed form](https://en.wikipedia.org/wiki/Closed-form_expression) solution to the [ordinary differential equation](https://en.wikipedia.org/wiki/Ordinary_differential_equation) defined by the "relative" expression.

When you use the "absolute" expression, you're simply plugging in timestamps into the closed form expression,
When you use the "relative" expression, you're actually performing [numerical integration](https://en.wikipedia.org/wiki/Numerical_integration).

In fields like physics, we *discover* the differential equation governing some natural process ("How does this change with x?"), but here we actually *choose* the differential equation ("How do we want this to change with x?") governing the system.

"Relative" and "absolute" ASERT are not different algorithms, they're the same mathematical object viewed through integration vs differentiation.

Similarly, the WTEMA is also an ordinary differential equation (ODE):

`target_{n+2} / target_{n+1} = 1 + (solvetime - targetBlockInterval) / tau` ,

but this ODE is inherently path-dependent, so no closed form solution exists.

Comparing the two ODEs:

| Property | Linear form (WTEMA) | Exponential form ("relative" ASERT) |
|----------|---------------------|-------------------------------------|
| Log-domain behavior | Nonlinear | **Linear** |
| Closed-form sum | No | **Yes** |
| Multiplicative consistency | No | **Yes** |
| Path independence | No | **Yes** (depends only on total time and block count) |

ASERT's ODE makes log-target a simple accumulator of timing errors, which is exactly what lets the absolute form exist, and what you want for a difficulty adjustment algorithm.

Multiplicative consistency practically means that in ASERT, a 5-min block will perfectly cancel a 15-min block: difficulty before and after the two blocks will be exactly the same. In WTEMA, this is not the case: a fast block followed by an equally slow block leaves difficulty slightly higher than before, causing systematic drift under volatility.

**Comment by A60AB5450353F40E** [comment]
>It will also be a problem if a coin changes the POW and needs to make a large sudden change to the difficulty. Both of these are easily solved by changing the baseline difficulty to the difficulty at the fork, but it's something to be aware of.

PS you can change target time the same way - scale the anchor to pretend it always had the new target time. I really appreciated ASERT's mathematical elegance when [I had to adapt it to changeable target block time](https://gitlab.com/0353F40E/fablous/-/blob/master/readme.md?ref_type=heads#difficulty-adjustment-algorithm-daa).

The proposal is to switch BCH's target time to 1 minute, and make later switches easier, so we introduce the "ticks" abstraction to specify chain height using uniform units (ideal milliseconds, one 10-min blocks is worth 600k "ticks", one 1-min block is worth 60k ticks).

But it would work for variable target block time chains like Nervos, too.

**Comment by zawy12** [comment]
I appreciate you pointing out that absolute and relative are the integral and differential.

As you probably know, WTEMA is almost exactly equal to relative ASERT for large tau (N in my lingo), and both suffer from nBits not being perfectly accurate which is alleviated in absolute ASERT.  At N=300 the difference is 0.0005% and there's a 300 * 2^(-15) = 1% possible error from nBits.  An additional error on a similar scale is the 2^x calculation. With 1 minute blocks N needs to be large and it would be be a big mess with relative in addition to the nBits error, so absolute is really really needed.  

You may know I believe monotonic timestamps is the most important thing coins aren't doing that they need to do (instead of MTP). If the code everywhere is good and correct (not subtly and disparately dependent) , the only adjustment necessary is to change MTP = 11 to MTP = 1.

You're interested in perfection, so I'll point out that 5 minute intervals occur more often than 15 minute intervals, so there's a little problem with ASERT. It's constantly trying to make difficulty too hard ("overshoots") and then has to correct itself as time goes on to get the right solvetime.  However, with large N the problem isn't noticeable.  For smaller N, the solution is to make the adjustment linearly-dependent on the probability of the expected solvetime which is based on the exponential distribution.  Let me go through the logic carefully: after we make an adjustment based on all prior data, that difficulty for that block is "our belief" or observation.  For the next block we have only 1 new piece of data, the solvetime (at the prior difficulty), and that piece should be the only thing necessary to adjust the prior difficulty. 

Relative ASERT is an EMA adjustment where 1-t/T is an error signal (it's t/T-1 if working in targets instead of difficulty). N is just a filter. The error signal based on the probability distribution of solvetimes is shown below. Experiment shows this perfectly gets the correct average solvetime down to N=1 where either ASERT will have a large error (30%?). 

It looks like I can't get a closed form for this method, so it's not practical due to the accumulation of errors if N > 1000. Also, if miners collude to set the timestamps all equal to target_time, the difficulty will get easier and easier. This would risk getting ahead of the FTL, so they would keep lowering the reported solvetime until blocks are being reported to have 0.693 * target_time and coming at 0.693 * target_time.  To solve this, I would like to see honest miners enforcing timestamps to within +/-10 seconds of block arrival unless an alternate chain gets ahead by 3 blocks. This prevents selfish mining with < 50% hashrate from being profitable which could be a big problem with 1 minute blocks. See my recent articles on this in trying to help Monero.  They didn't want to use this idea.

<img width="662" height="200" alt="Image" src="https://user-images.githubusercontent.com/18004719/163170221-10b8d914-0053-4c93-bcc9-a1b8cad07858.png" />

**Comment by A60AB5450353F40E** [comment]
>You may know I believe monotonic timestamps is the most important thing coins aren't doing that they need to do

In any case the clock can be frozen for a while (advancing +1s every 6 blocks, so +1s every hour for 10-min target or +10s every hour for 1-min target; monotonic would only slightly speed it up to +1s/block while frozen).
But with monotonic it'd get reset to current time more frequently (whenever a honest miner finds a block vs whenever they find 6/11) so frozen periods would be shorter and would lag less behind real time. I'm still studying what this means for DAA, can't comment much now. I do like the idea of monotonic timestamps.

>I'll point out that 5 minute intervals occur more often than 15 minute intervals

Right, the exponential distribution is asymmetric. But does this bias matter practically at the `tau` values for ~2-day half-life? We'd maintain the same for 1-minute blocks. My understanding is that with large `tau` the asymmetry averages out, and the "overshoot" you describe becomes negligible.

Maybe this would be more relevant if someone tried real-time-targeting + small `tau`, what kind of mining incentives would that create? With RTT, wouldn't miners just try stick as close to FTL as possible, else they'd give the next miner the opportunity to mine at reduced difficulty.

>suffer from nBits not being perfectly accurate

BTW, is there even a reason why nBits have to be part of the header? After all, they can be derived from just the header timestamps. The DAA could maintain state at higher precision. Instead of "loading" it from the last header, it could just compute it from the whole header chain. Same how our [adaptive blocksize limit algorithm](https://gitlab.com/0353F40E/ebaa/-/blob/main/README.md#technical-description) maintains internal state and uses it to spit out the blocksize limit. It needs the whole history of block sizes to compute state for current block.

**Comment by dgenr8** [comment]
> Maybe this would be more relevant if someone tried real-time-targeting + small `tau`

By not including such a parameter, RTT defines tau to be 1.  Every block is an experiment that completely resets the anchor for the next block.

> With RTT, wouldn't miners just try stick as close to FTL as possible?

Yes. This is why RTT rejects any block claiming to be from the future.

**Comment by zawy12** [comment]
I meant to mention RTT at the end to point out that's what we really want.  Monotonic timestamps and enforcing timestamps with accuracy << block time is necessary to get a valid consensus to less than 1 block. Selfish mining attacks are possible with 33% hashrate when the 2nd rule is broken as Lamport showed in 1978. The timestamps need to be multiples of the allowed timestamp error (2x the individually-allowed clock error). This allows validation to simply be "did timestamp match local time when the block arrived" (I'm ignoring a propagation delay complication). But more importantly it prevents an attacker from carefully selecting timestamps and releasing the block to "fall on top of" the point of decision, splitting the network in half in deciding if the timestamp was accepted or not.  This was also in the 1978 paper and it confused me for years until I realized <strike>following</strike> it fixed this possible attack on RTT. To clarify, the allowed multiples of timestamps must be spaced apart more in time than the "typically 99% max" propagation delay plus 2x the allowed local clock error. Then if the timestamp doesn't match the timestamp that local time would have selected when it arrives, it's rejected.

Monero clones didn't include a difficulty in the header. They calculated it block by block when validating from the beginning. That was a bad idea, but with absolute ASERT, I guess you don't need it.

> My understanding is that with large tau the asymmetry averages out, and the "overshoot" you describe becomes negligible.

CDF-EMA responds faster with better stability. The "3" in the formula above gives it equal stability to ASERT for a given value of N and a lot faster response. I determined the 3 by experiment.  It should be "2" to compare them more directly (the 0.5 becomes a 1 where ASERT has a 1). This would give it better stability and slightly better speed of response to hashrate changes. 

You could use absolute ASERT and RTT to make an absolute version of CDF-EMA and not have an nBITs.  This would also fix the "0.693" hack on CDF-EMA that I described above. Using RTT is optional, but it fits in naturally: You calculate a "ghost" difficulty target based on absolute ASERT for the tip you're working on. That ghost difficulty, the tip's timestamp, and your local time (RTT timestamp) while mining are the inputs to the CDF-EMA. The N is the same for both DAAs. Use 2 in place of 3 in CDF-EMA.



$$
\text{Target}_{n+1} = \text{absolute-ASERT-target}_{n} \cdot e^{\frac{\left( e^{-\frac{t}{T}} - 0.5 \right) \cdot 2}{N}}
$$

t = solvetime from n to n+1.
T = target block time.

This doesn't harm the effectiveness of CDF-EMA and it ensures you get the correct average block time even if a collusion of miners is trying to throw the CDF-EMA off (get an easier difficulty for faster blocks) by not following the exponential distribution. I think this is neat.

Combining the two like this prevents the CDF-EMA from using the precise target from the previous block as it should. The effect is to slow the speed of response to hashrate changes down. 

The two combine into this formula:



$$
\text{Target}_{n+1} = \text{Target}_0 \cdot e^{\frac{\frac{t_n - t_0}{T(n - n_0)} - 1}{N}} \cdot e^{\frac{2 \cdot e^{-\frac{t_{n+1} - t_n}{T} - 1}}{N}}
$$


$$
\text{Target}_{n+1} = \text{Target}_0 \cdot \left( \exp^\left({\frac{t_n - t_0}{(n - n_0)T} + 2e^{-\frac{t_{n+1} - t_n}{T}} - 2} \right)\right)^{\frac{1}{N}}
$$

To interpret this, the first two terms in the exponent should have been 1, so their sum should be 2 and after subtracting the 2, we have e^0 = 1 = no adjustment.  The first term (absolute ASERT) is to correct for long-term error and the 2nd (CDF-EMA) is to correct for the error in current RTT solvetime.

**Comment by zawy12** [comment]
There's a small error in relative ASERT that's less of an error in absolute ASERT. Where I have n - n_0 in the last equation should be n - n_0 - 1.

DAAs look back a specific number of blocks to measure the solvetime which means the Erlang distribution applies which means an (N-1)/N correction is needed. If we do the opposite and look back a fixed amount of time and count blocks, the Poisson applies. **The (N-1)/N adjustment allows an Erlang observation to simulate a Poisson observation.**  The adjustment allows us to pretend we counted N blocks after deciding to look back in time an amount equal to the timespan. 

Consider the memoryless property. In some sense we should expect there will be 1 solvetime before the 1st block in N blocks and 1 solvetime after the last, so there is "one extra solvetime" than the number of blocks. 

[I convinced PW of this at Delving](https://delvingbitcoin.org/t/correcting-the-error-in-getnetworkhashrateps/1745/3) in correcting how we measure hashrate. Hashrate is not:

sum_difficulty / sum_solvetimes  

when we look back N blocks.  It's: 

sum_difficulty / (sum_solvetimes + avg_solvetime)
or
sum_difficulty / sum_solvetimess  * (N-1) / N

In both hashrate and DAAs, we can make a Poisson (time) observation instead of an Erlang observation (blocks) like this:

hashrate = sum_difficulties / fixed_timespan

This means we choose a timespan, then count the blocks in it. You need to choose a random starting point of the timespan. For example, midnight every night or whenever we randomly decide to estimate hashrate. I assume a block solvetime is a random starting time, so I can use a block's timestamp provided I don't count that block as part of N, thereby it replaces the need for the (N-1)/N adjustment.  An RTT's local time seems to serve as a random starting time. That time is not used unless the block is found, which should be random. 

Looking back at ASERT, all this means the n-n_0 in my denominator in the last equation needs to be n - n_0 - 1, or _maybe_ I can use the current time, i.e. t_{n+1} instead of t_n. My hesitation in using current time is that it's a double-adjustment based on the current time. That option is appealing because it removes the need for the prior block's timestamp  Notice absolute ASERT will have very little error compared to relative ASERT if this adjustment isn't applied because N in the (N-1)/N for absolute ASERT is very large.

CDF-EMA here is like P (proportional) in PID controller ("present time" adjustment).  ASERT is more like "integral" (sum of past measurements) but the EMA aspect makes it also like P.  My experience is that the logic "last block had best estimate" mean P is the best that can be done.  D would measure the derivative. I made several attempts and any PID efforts were fruitless. The main problems is that there's no momentum like there are with systems with momentum that PID is meant for. A strictly integral method would be to apply an SMA in place of ASERT.

I'm doubtful the combination is better than CDF-EMA alone except being able to remove the need to look at the prior block and correcting for potential manipulation.

**Comment by A60AB5450353F40E** [comment]
>In some sense we should expect there will be 1 solvetime before the 1st block in N blocks and 1 solvetime after the last, so there is "one extra solvetime" than the number of blocks.

The ["hitchiker's paradox"](https://r6.ca/blog/20180225T160548Z.html)

To get the intuition for this: Poisson means that shorter times will be more frequent and all intervals together will produce the target average, but, the longer intervals will occupy **more of the timeline**. I vibe-coded a little game to visualize this: https://fablous-18ff29.gitlab.io/?d=2026-02-02

When you randomly open a block explorer you pick a random point on the timeline, and chances are higher you will land inside one of the longer intervals, because they occupy more space, and your interval duration will be given by Erlang. But where inside the interval will you land? Wait time until end of interval will be given by Poisson, as will the elapsed time since start of interval.

**Comment by A60AB5450353F40E** [comment]
>With 1 minute blocks N needs to be large and it would be be a big mess with relative in addition to the nBits error, so absolute is really really needed.

Yeah, [my analysis of timewarp resistance under faster blocks](https://gist.github.com/A60AB5450353F40E/1489a859dbe6a26926d37e2f560104ae) shows that with "absolute" ASERT, `tau` still at 2 days, MTP still defined over 11-block window (11 minutes with 1-min blocks), and FTL still at +2h -- nothing fundamentally changes for BCH security if it should change target time from 10 minutes to 1 minute.

Absolute ASERT is robust to timestamp manipulations. What is borrowed must be paid back, or it can be rolled over indefinitely if you keep timestamps stuck at FTL. But with big `tau` that will be just a few % of extra profitability lasting couple of weeks. Then, everyone will be stuck at FTL unless they're ready to "pay back" what they borrowed (increased difficulty) in order to restore the clock.

ASERT is like one big rubber band stretching all the way to anchor block. Each block is like inserting a new pre-stretched piece of rubber -- and if it's not pre-stretched to the same stretch then the whole rubber band relaxes or contracts, so the next piece of rubber is allowed to come with less or more pre-stretch in order to restore stretch to target tension.

>Monero clones didn't include a difficulty in the header. They calculated it block by block when validating from the beginning. That was a bad idea

Why was it a bad idea? Is it because of not being able to discard invalid blocks on their own without chain context? When target is part of header, you have all you need to evaluate whether it has enough PoW. But maybe these concerns could be separated: lossy bits for anti-DoS PoW check, but computed full precision for final check.

**Comment by zawy12** [comment]
I think Monero clones would sometimes not be able to make desired changes because of not having the difficulty in the header. I can't remember exactly what the problem.  I am not saying that ASERT has a problem because you don't have to recalculate the whole chain to figure out current difficulty.

**Comment by A60AB5450353F40E** [comment]
Yeah, but if you wanted higher precision for EWMA then you could use full precision for calculations and block acceptance and only use `nBits` to store rounded up target as commitment. Anti-DoS could verify hash against `nBits` as first pass, and online consensus would validate it against full precision target derived from whole chain of headers, call it `fullBits`. The `fullBits` would always be a slightly harder target than `nBits`.

>You calculate a "ghost" difficulty target based on absolute ASERT for the tip you're working on. That ghost difficulty, the tip's timestamp, and your local time (RTT timestamp) while mining are the inputs to the CDF-EMA.

This sounds really promising. What does "CDF" stand for? What's the optimal strategy for mining? Should miners pause mining for a while until the RTT part drops difficulty? I guess RTT would reduce variance, too, less blocks with short solvetimes (miners pause for harder difficulty), but also less blocks with long solvetimes (miners race for easier difficulty).

**Comment by zawy12** [comment]
CDF mean "cumulative distribution function".  My CDF-EMA adjustment works for any system that needs to be controlled if you know the CDF of the error-signal. In our case it's the exponential CDF. 

Yes, miners could collude to keep their machines off in an RTT until say 1 minute before a 10-minute block time. This would cause the difficulty to get lower and lower in which case they would need to be more precise in their clocks to make sure no one cheats on the collusion. If anyone cheats, they ignore the block. Again, they can see cheating if the timestamp isn't close enough to block arrival time.  But this can be used just as easily in a regular block chain to remove 90% of the OPEX expenditures.  The wasted electricity (OPEX) isn't what provides bitcoin security, it's the proof of CAPEX in the equipment.  Not wanting to devalue their CAPEX by hurting the coin's operation is the only thing that secures bitcoin.  This scheme also allows old equipment to be used for longer which increases security.

This could be achieved more directly by alternating a VDF and PoW. See PoWVDF

**Comment by dgenr8** [comment]
> I made several attempts and any PID efforts were fruitless.

Mining is stochastic, and retargeting directly controls the parameter, with the goal of producing blocks fairly and consistently regardless of how hashrate changes.

Analyzing samples using time-series techniques like a PID controller or even a moving average is unnecessary. Correcting for fast or slow production in the recent or distant past is not helpful.  Changing difficulty slowly is not important.

Attempts to graft RTT onto other systems tend to ignore that by design, it strongly and symmetrically damps its own response, rather than massaging its inputs.

**Comment by A60AB5450353F40E** [comment]
> But this can be used just as easily in a regular block chain to remove 90% of the OPEX expenditures.

How so? Sure, Bitcoin's original DAA can be abused to keep difficulty low, but ASERT can't. If with ASERT (tau = 2 days) everyone sets their timestamp to FTL then they get one-time difficulty reduction to -2.85% from baseline which lasts for a couple weeks and then it's back to baseline difficulty and stuck mining at FTL. Restoring the clock would do the opposite, bump difficulty to +2.93% from baseline, and then they'd have to suck it up until it decays back to baseline. What has been borrowed with FTL clock must be given back to restore the clock.

**Comment by A60AB5450353F40E** [comment]
Hey @zawy12 as I mentioned above, we're changing BCH's target block time to 1 minute ([proposal](https://gitlab.com/0353F40E/fablous), [discussion](https://bitcoincashresearch.org/t/chip-2025-03-faster-blocks-for-bitcoin-cash/1513)). In [their review](https://gitlab.com/GeneralProtocols/research/chips/-/blob/master/statements/CHIP-2025-03-Faster-Blocks-for-Bitcoin-Cash_Statement-2026-09-20.md), General Protocols (GP) has named you and @dgenr8 as one of external experts whose feedback would be valuable.

>While there are many experts that might not respond or that might not consider themselves stakeholders, it is important that we do not dismiss or ignore their contributions or expertise. Therefore, the following people currently not listed in the [CHIPs stakeholder list](https://gitlab.com/0353F40E/fablous/-/blob/master/stakeholders.md) should at least be asked what their opinions are on the changes:

So, if either of you is willing to make some time and give some feedback, maybe have a look at the [DAA section](https://gitlab.com/0353F40E/fablous/-/blob/master/readme.md?ref_type=heads#difficulty-adjustment-algorithm-daa) and related ["ticks" abstraction](https://gitlab.com/0353F40E/fablous/-/blob/master/readme.md?ref_type=heads#blockchain-height-abstraction) which replaces raw height in formulas and all height-sensitive consensus rules. This way, we can in the future change block time again without having to touch all this height-sensitive code ever again. While working on this, I really appreciated ASERT's mathematical elegance - it was easy to re-express in ticks to support changeable target block time.

**Comment by dgenr8** [comment]
While the ASERT response to a hashrate change may be quantitatively stable vs. pre-fork, being evaluated 10x more often decreases BCH reaction time to the change, giving it a further advantage against BTC.

**Comment by zawy12** [comment]
Jonathan Toomin might be willing to help. Your future time limit needs to be cut from 7200 to 720 if ASERT's half-life is still 288 blocks when you go to 60 seconds.  288 will be fine.  Some accuracy in block time will be lost if you let it go to 2880, maybe 5%.   It's easier and just as good to use EMA instead:

next_target = target + target * t // (60 * 415) - target // 415

BCH uses absolute ASERT instead of relative ASERT, so the starting height will need to be changed, or some other finagling. 

Getting rid of the MTP trash by using monotonic would be nice.

**Comment by A60AB5450353F40E** [comment]
Thanks dgenr8, yeah reaction time is a nice benefit, also some big hashpower drop becomes more bearable and requires less PoW to "unstick" because ASERT only needs 1 sample. I know RTT solves this, but we aren't ready for that conversation (yet) :)

And thanks zawy12, we're keeping the half-life fixed in wall-clock terms so it gets scaled to 2880 in block count, and we're not touching FTL or switching to EMA since I don't want to burden the proposal with out of scope changes.

>BCH uses absolute ASERT instead of relative ASERT, so the starting height will need to be changed, or some other finagling.

Yeah, scaling the anchor block's difficulty and replacing the +1 height term with block's target time does the trick. You foreshadowed this in your opening post above:

>It will also be a problem if a coin changes the POW and needs to make a large sudden change to the difficulty. Both of these are easily solved by changing the baseline difficulty to the difficulty at the fork, but it's something to be aware of.

The MTP trash count stays at 11, I'd like to throw out this trash one day and have it monotonic, but that too is out of scope. At least now it will get closer to real time (~6 min behind).

#### Block time accuracy

>Some accuracy in block time will be lost if you let it go to 2880, maybe 5%.

What do you mean by this? Is it about this that you wrote above:

>People forget that ASERT gets behind in the number of blocks if hashrate continually increases because it has perfect solvetime if hashrate is oscillating around some average. Getting behind is how it increases difficulty.

and 

>BTC has had about 6% faster solvetime than desired due to hashrate increasing faster than the DA. What would BCH's avg solvetime have been if it had used ASERT with half life 288 from the beginning? Difficulty is 350E9 and it started at D=1. The emission schedule wanted about 605,000 blocks emitted by now instead of 640,000 (5.7% too fast).

But that's the status quo, right? Scaling it to 2880 just preserves the status quo in that regard.

BTW, in 2017 BCH's emission schedule went ahead by 131k blocks thanks to the horrible EDAA we had for those few months, so idk how you got the 640k vs 605k numbers.

**Comment by zawy12** [comment]
No, neither of those are the source of the error I had in mind.  There is a little error in the 2^x calculation and that error will be 10x worse now.  nBits has an error up to 2^(-15).  But I was forgetting that one of the arguing points for absolute ASERT was that it does not have the error like relative ASERT and EMA.  But in the first set of 2880 blocks it will have 3% to 10% error in the block rate even if the difficulty starts out perfect for the hashrate. The that would be cut in half in the 2nd set, to 1/3 in the 3rd.  

Concerning BCH blocks emitted, I believe I was referring to the interim DAA after the disaster that I believe was a 144-block SMA.  I think deadalnix got that one from one of my posts recommending it, but before it went live I was pushing for some changes.

**Comment by A60AB5450353F40E** [comment]
Are you assuming we reset the anchor to 10->1 min activation block? Because we're not. We're still anchoring it in the old anchor, and simply scaling the old anchor's target 10x:

    next_target = anchor_target * (anchor_ideal_block_time / next_ideal_block_time) *
                    2**((time_delta - (tick_delta + anchor_ideal_block_time)) / halflife)

This is effectively pretending we had 1-min blocks all the way since the old anchor, so my intuition is that this error period would be far behind. I will check better, this is just my 1st pass reading of the issue.

**Comment by zawy12** [comment]
Right, I forgot you had just told me you're going to use the old target, so the error shouldn't apply.  

Your documentation didn't define tick_delta and anchor_ideal_block_time. 

I will just think out loud here. 

```tick_delta = "desired time since time_delta began in 60 increments"  = 600 * (H - H_a) + 60 * (h-H)```

H  = is one block before the 1st 60 s block is mined.
H_a  = height of block that is 1 before anchor block (to match tick_delta with time_delta).
h = height of the tip, not the current block we're mining ( to match tick_delta with time_delta)

Now let me change a variable to match what you're probably using:
G_a = anchor height

```tick_delta = 600 * (H - (G_a - 1)) + 60 * (h-H)```

Defining it this way prevents the need for anchor_ideal_block_time. Removing parenthesis helps the fractional math if it weren't an even division
```
next_target = anchor_target * anchor_ideal_block_time / next_ideal_block_time * 
                2**((time_delta - tick_delta) / halflife)
```
Your half-life is 288, so everything I said earlier about larger error doesn't apply even if was relative ASERT. Again, it's because I wasn't reading and thinking carefully.

**Comment by A60AB5450353F40E** [comment]
Sorry, the current tick is cumulative sum of all past target times since genesis. The `ticks_delta` is difference between current tick and anchor tick. It abstracts the measure of chain progression since anchor from block count to idealized seconds. Each 10-min blocks contributes exactly 600 ticks, and each 1-min block contributes exactly 60 ticks. So we can measure the "stretch" all the way back to the old 600-tick anchor at height 661,647.

This ASERT formulation could be used on dynamic target time blockchains like Nervos, too.

So it's not "desired time since time_delta began in 60 increments" but "desired time since anchor", and the anchor could be the genesis block for new chains.

Yeah, half-life stays 172800 seconds (288 x 10-min or 2880 x 1-min). The part in the exponent is not changed at all. We simply scale the result to current target time after doing the exponent math on schedule deviation.

**Comment by zawy12** [comment]
Oh, right, me saying "in 60 second increments" was wrong compared to my math. My math should be the same as yours.  I just included the anchor_ideal_block_time that's in your exponent into tick_delta so that tick_delta is the same number of blocks as time_delta, as my definitions said.
